### PR TITLE
[UI][Grid] Added grid based pagination configuration

### PR DESCRIPTION
**Pull request by NeverResponse**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | yes |
| BC breaks?      | yes |
| Related tickets | - |
| License         | MIT |

The basic idea is to provide a fast way of setting how many items are visible on given grid's index.
So etc. the following:

``` yaml
grids: 
  sylius_admin_product:
    ...
    limits: [20, 10, 30]
    ...
```
will result in **20** items visible on the page by default, and will set **10** and **20** as choices.

If the `limits` parameter of grid's configuration is omitted, the default comes in, setting limits to
`[10, 25, 50]`.

Disabling the default limit can be done by passing either a null or an empty array so:

``` yaml
grids: 
  sylius_admin_product:
    ...
    limits: [ ]
    ...
```
or

``` yaml
grids: 
  sylius_admin_product:
    ...
    limits: ~
    ...
```

will do the trick, and fallback to 10 visible per page with no select available on page.
The limit selector is hidden in two cases: 
 - there are less items available than the default value (same thing happens to the paginator's page select)
 - either one, or none limits configured

And since a picture's worth a thousand words:

- default configuration, 10 per page

screen shot 2017-02-08 at 13 25 58

 - 50 per page

screen shot 2017-02-08 at 13 26 36

### CODE REVIEW

File: src/Sylius/Bundle/AdminBundle/Resources/config/grids/customer.yml [hunk 1]
```diff
@@ -7,6 +7,7 @@ sylius_grid:
                     class: "%sylius.model.customer.class%"
             sorting:
                 createdAt: desc
+            pagination: [10, 25, 50]
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/AdminBundle/Resources/config/grids/customer.yml]: I think that is a sensible default, wdyt?

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourcesCollectionProvider.php [hunk 2]
```diff
@@ -48,12 +48,19 @@ public function __construct(ResourcesResolverInterface $resourcesResolver, Pager
...
+            $paginations = $resources->getDefinition()->getPagination();
+        } else {
+            $paginator = $resources;
+        }
 
         if ($paginator instanceof Pagerfanta) {
             $request = $requestConfiguration->getRequest();
-            $paginator->setMaxPerPage($requestConfiguration->getPaginationMaxPerPage());
+
+            $paginator->setMaxPerPage($request->query->get('paginate', reset($paginations) ?: $requestConfiguration->getPaginationMaxPerPage()));
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourcesCollectionProvider.php]: ``paginate`` -> ``limit``? Thoughts anyone? :)

Comment (steffenbrem) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourcesCollectionProvider.php]: Always was confused why this was different for paginated resources. Would like to see consistency in using `limit` for paginated resources too.

File: src/Sylius/Component/Grid/spec/Definition/ArrayToDefinitionConverterSpec.php [hunk 3]
```diff
@@ -70,6 +72,11 @@ function it_converts_an_array_to_grid_definition()
             'sorting' => [
                 'code' => 'desc',
             ],
+            'pagination' => [
```
Comment (GSadee) [inline src/Sylius/Component/Grid/spec/Definition/ArrayToDefinitionConverterSpec.php]: Why not `'pagination' => [10, 25, 50],` ?

File: src/Sylius/Bundle/GridBundle/Tests/DependencyInjection/ConfigurationTest.php [hunk 4]
```diff
@@ -182,6 +204,100 @@ public function its_sorting_order_can_be_only_ascending_or_descending()
...
+                ]],
+                ['grids' => [
+                    'sylius_admin_tax_category' => [
+                        'limits' => [6, 12, 24],
+                    ],
+                ]],
+            ],
+            ['grids' => [
+                'sylius_admin_tax_category' => [
+                    'limits' => [6, 12 ,24],
```
Comment (pamil) [inline src/Sylius/Bundle/GridBundle/Tests/DependencyInjection/ConfigurationTest.php]: `12 ,24` -> `12, 24`

File: src/Sylius/Bundle/GridBundle/Tests/DependencyInjection/ConfigurationTest.php [hunk 5]
```diff
@@ -182,6 +204,100 @@ public function its_sorting_order_can_be_only_ascending_or_descending()
...
+                'sylius_admin_tax_category' => [
+                    'limits' => [10, 25, 'surprise!']
+                ]
+            ]
+        ]]);
+
+        $this->assertConfigurationIsInvalid([[
+            'grids' => [
+                'sylius_admin_tax_category' => [
+                    'limits' => ['hello', 'mister', 'pug']
```
Comment (pamil) [inline src/Sylius/Bundle/GridBundle/Tests/DependencyInjection/ConfigurationTest.php]: Do we need this one if we got the one with strings above?

File: src/Sylius/Bundle/ResourceBundle/spec/Controller/ResourcesCollectionProviderSpec.php [hunk 6]
```diff
@@ -76,6 +77,7 @@ function it_handles_Pagerfanta(
 
         $requestConfiguration->getRequest()->willReturn($request);
         $request->query = $queryParameters;
```
Comment (pamil) [inline src/Sylius/Bundle/ResourceBundle/spec/Controller/ResourcesCollectionProviderSpec.php]: Mocking `ParameterBag` always hurts :(

File: docs/bundles/SyliusGridBundle/your_first_grid.rst [hunk 7]
```diff
@@ -188,6 +188,30 @@ You can define by which field you want the grid to be sorted and how.
                     name: asc
                     ...
 
+Pagination
+----------
+
+You can limit how many items are visible on each page by providing an array of integers into the ``limits`` parameter.
+The first element of the array will be treated as the default, so by configuring:
+
+
```
Comment (Zales0123) [inline docs/bundles/SyliusGridBundle/your_first_grid.rst]: Redundant line.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/Index/_pagination.html.twig [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+<div class="ui right floated small header">
+    <div class="content">
+        <div class="ui inline dropdown sylius-paginate">
+            {{ 'sylius.ui.show'|trans }} {{ resources.data.maxPerPage }}
+            <i class="dropdown icon"></i>
+            <div class="menu">
+                {% for limit in paginationLimits if limit != resources.data.maxPerPage %}
+                    {% set path = path(app.request.attributes.get('_route'), app.request.attributes.get('_route_params')|merge({'limit': limit})) %}
+
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Product/Index/_pagination.html.twig]: Again.

**Comment by sweoggy** [comment]
Great addition! I was just wondering if we could also display the entry numbers of currently displayed items out of the total number of entries? Just like https://datatables.net/ does it:

Example: "Showing 11 to 20 of 57 entries" on second page of a table with 10 entries on each page with a total of 57 entries.

**Comment by NeverResponse** [comment]
@sweoggy Think it's enough as is; for the entry point at least.
Customizing it shouldn't be a problem if someone actually wants to add the counter you're talking about.
Since it'd be a slight UI change, with no BC breaks, i see no problem with introducing it a bit later on, provided there's enough demand for it to be in the core.

**Comment by pjedrzejewski** [comment]
Thanks Jasiek!

**Pull request merged**